src/Corax/Queries/SuggestionTermProvider.cs

```diff
@@ -0,0 +1,140 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+using Corax.Pipeline;
+using Corax.Utils;
+using Sparrow;
+using Sparrow.Collections;
+using Sparrow.Server;
+using Voron;
+using Voron.Data.CompactTrees;
+
+namespace Corax.Queries
+{
+    public interface IRawTermProvider
+    {
+        void Next(ref Span<byte> terms, ref Span<Token> tokens);
+    }
+
+
+    public unsafe struct SuggestionTermProvider : IRawTermProvider
+    {
+
+        public const int DefaultNGramSize = 4;
+
+        private readonly IndexSearcher _searcher;
+        private readonly int _fieldId;
+        private readonly Slice _term;
+        private readonly Analyzer _analyzer;
+        private readonly int _take;
+        private readonly delegate*<ref SuggestionTermProvider, ref Span<byte>, ref Span<Token>, void> _nextFunc;
+
+        public SuggestionTermProvider(
+            IndexSearcher searcher, int fieldId, 
+            Slice term, Analyzer analyzer, int take,
+            delegate*<ref SuggestionTermProvider, ref Span<byte>, ref Span<Token>, void> nextFunc)
+        {
+            _searcher = searcher;
+            _fieldId = fieldId;
+            _term = term;
+            _analyzer = analyzer;
+            _take = take;
+            _nextFunc = nextFunc;
+        }
+
+        private struct ScoreComparer : IComparer<(Slice, int)>
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public int Compare((Slice, int) x, (Slice, int) y)
+            {
+                return x.Item2 - y.Item2;
+            }
+        }
+
+        public static SuggestionTermProvider YieldFromNGram(IndexSearcher searcher, int fieldId, Slice term, Analyzer analyzer, int take = -1)
+        {
+            static void Next(ref SuggestionTermProvider provider, ref Span<byte> terms, ref Span<Token> tokens)
+            {
+                int fieldId = provider._fieldId;
+                var allocator = provider._searcher.Allocator;
+                var term = provider._term;                
+
+                Slice.From(allocator, $"__Suggestion_{fieldId}", out var treeName);
+                var tree = provider._searcher.Transaction.CompactTreeFor(treeName);
+
+                var keys = NGrams.Generate(allocator, DefaultNGramSize, term.AsSpan(), out int keysCount).ToReadOnlySpan();
+                int keySizes = keys.Length / keysCount;
+
+                var values = new FastList<(Slice, int)>();
+                var dictionary = new Dictionary<Slice, int>();                
```

Comment:
Worth caching this?

Reply:
Probably. But wont go there yet until I know that what I am implementing is the right thing to do... usually handle of that in the last pass over it.